HeapStatData with a distinct allocation mechanism for RawStatData

jmarantz opened the pull request:
*Description*:
>This PR sets the stage for a memory-friendly stat-allocator for use with hot-restart disabled.

- hot-restart stats are still in RawStatData, non-hot-restart stats are in HeapStatData. Previously they were in HeapRawStatData which derived from RawStatData, whereas HeapStatData is an alternate impl, not derived from RawStatData.
- allocators are now responsible for truncation, and require StatsOptions to be passed on construction
- IsolatedStore now truncates at 1M chars, rather than special-casing no-truncation. This is now tested.
- some tests moved around as there is no allocator for RawStatData in stats_impl.cc anymore; it's
   in hot_restart_impl.cc
- factored out a mocked testing stats-allocator from 3 different tests into a common one, changing
it from being multi-inherited with the testing framework to being instantiated. This was needed because stat-allocators now require StatOptions in their constructor.

This is another step toward addressing https://github.com/envoyproxy/envoy/issues/3585

*Risk Level*: Medium
*Testing*: //test/...
*Docs Changes*: N/A
*Release Notes*: N/A

mattklein123 commented:
@dnoe please take a first pass on this

jmarantz commented:
@mattklein123 @dnoe I mainly issued this to share the ideas here with @ambuc  -- current thinking is we want his https://github.com/envoyproxy/envoy/pull/3629 to go in first, I think.

jmarantz commented:
does pinging this re-open the PR?  Now that #3629 is in I want to resolve conflicts & get this rolling.

jmarantz commented:
I think this is ready for a look, but note that I may have regressed the behavior of IsolatedStore. See comments at the relevant line.

jmarantz commented on source/common/stats/stats_impl.h:
```diff
@@ -443,8 +495,16 @@ template <class Base> class IsolatedStatsCache {
+  // https://github.com/envoyproxy/envoy/pull/3629, which removed the shadowing
+  // of option size in a static.
+
+  /* explicit */ IsolatedStoreImpl(/*Stats::StatsOptions& stats_options*/)
```
Note behavior regression here:

- prior to #3629 : truncated the same as other stats due to shared static max length
- after #3629 : not truncated at all (AFAICT)
- with this PR : truncated to default length (127 chars).

If someone (@ambuc @mattklein123 @mrice32) can share the desired behavior (not truncated at all vs truncated with the other stats) I will fix.

ambuc replied:
I think this comes down to the question of whether or not we expect an IsolatedStore to have stats which are interoperable with other stats. My instinct is no, due to its isolated nature -- in which case this behavior is not something anyone relies upon. Thoughts?

jmarantz replied:
Per discussion, I force the truncation at 1M bytes, so effectively we are not truncating in this PR.

jmarantz commented on test/common/stats/stats_impl_test.cc:
```diff
@@ -478,40 +478,26 @@ TEST(TagProducerTest, CheckConstructor) {
 }
 
-// Check consistency of internal stat representation
-TEST(RawStatDataTest, Consistency) {
```
Note: moved to hot_restart_impl_test.cc, which allocates RawStatData*. stats_impl.cc no longer provides an allocator that allocates RawStatData*.

jmarantz replied:
This test -- still in hot_restart_impl_test.cc -- no longer covers truncation. That's covered in a separate new test in thread_local_store_test.cc, which is solely responsible for truncation.

This test's sole purpose is too make sure that the bit-pattern of a shared-memory stat allocation remains unchanged across releases, unless we also update SharedMemory::VERSION.

jmarantz commented on test/test_common/utility.h:
```diff
@@ -371,6 +374,27 @@ class TestAllocator : public RawStatDataAllocator {
   std::unordered_map<std::string, CSmartPtr<RawStatData, freeAdapter>> stats_;
 };
 
+class MockedTestAllocator : public RawStatDataAllocator {
```
Note: this mocked allocator was previously repeated 3x in the codebase, all as part of multiple inheritance with a test class. What motivated the refactor in this PR, though, was the need to pass a StatsOptions& to the constructor, which is annoying to do for a test-class.

If desired I could split that out as a separate PR.

ambuc commented on source/common/stats/stats_impl.cc:
```diff
@@ -135,35 +135,47 @@ bool TagExtractorImpl::extractTag(const std::string& stat_name, std::vector<Tag>
-  if (data == nullptr) {
-    throw std::bad_alloc();
-  }
-  data->checkAndInit(name, num_bytes_to_allocate);
```
If we're getting rid of checkAndInit, is there other cleanup in the regular RawStatData struct we can do now, now that it's been decoupled from HeapStatData? Maybe RawStatData, if only for blocked / shared memory, can be less general and more opinionated about its internal memory representation.

jmarantz replied:
I don't think we're getting rid of checkAndInit. It is still needed in the shm case. I don't think anything gets easier by knowing it's always shm.

jmarantz replied:
correction...checkAndInit is just merged with initialize() now.

ambuc commented on source/common/stats/stats_impl.h:
```diff
@@ -500,11 +560,11 @@ class IsolatedStoreImpl : public Store {
   };
 
-  HeapRawStatDataAllocator alloc_;
+  StatsOptionsImpl stats_options_;
```
nit: `const`?

jmarantz replied:
no longer const-able; I need to update it after allocating it, in order to set the truncation limit high.

ambuc commented on source/server/hot_restart_impl.cc:
```diff
@@ -144,11 +145,7 @@ HotRestartImpl::HotRestartImpl(Options& options)
-    key.remove_suffix(key.size() - options_.statsOptions().maxNameLength());
-  }
-  auto value_created = stats_set_->insert(key);
+  auto value_created = stats_set_->insert(truncateStatName(name));
```
This is really nice -- maybe we could use this for the explicit `name.substr()` in `ThreadLocalStoreImpl::ScopeImpl::safeMakeStat` ?

jmarantz replied:
Yeah I don't think we need that anymore; it's done in the allocator.  Removing that (along with your TODO).

jmarantz commented:
ping: this is assigned do @dnoe but it probably should be reassigned as he is currently OOO.

htuch commented:
@jmarantz I'll handle this PR. Can you provide a description of each of the refactorings/changes you've made and the motivation behind them? Ideally also in the PR description, as it's not clear really what's happening there other than this is a stepping stone.

mrice32 commented:
@htuch @jmarantz I can take a first pass.  Sorry, I meant to go through it yesterday.

mrice32 commented on source/common/stats/stats_impl.cc:
```diff
@@ -135,35 +135,47 @@ bool TagExtractorImpl::extractTag(const std::string& stat_name, std::vector<Tag>
+HeapStatData::HeapStatData(absl::string_view key) : name_(key.data(), key.size()) {}
 
+HeapStatData* HeapStatDataAllocator::alloc(const std::string& name) {
+  auto data = std::make_unique<HeapStatData>(truncateStatName(name));
```
I didn't see this behavior change documented in the PR description - are we now going back to truncating in non hot restart mode?  I think we should probably think carefully about what we want the user-facing behavior to look like long-term before changing the behavior again to avoid churn for something that is at least nominally user-facing.  As @mattklein123 has noted in the past, this behavior has become quite complex, so I think it's worth some thought to simplify the user-facing surface and document the behavior.

Google internally and (I assume) most other users of Envoy in non hot restart mode have no reason to want stats to be truncated.  I think it's totally okay to allow truncation in that mode, but if we do, I think it would be better to allow it to be configurable via the command line.  We could try to incorporate this optionality into the existing command line parameter somehow (for instance, allow users to pass in 0 for no truncation).  WDYT?

Side note: I'm less concerned about the isolated stats store as those stats are not exported to or queryable by the user unless made available through other means.  We could add a way to pass in stats options from whatever code is creating it, but I see no reason why it needs to conform to the parameters used for the primary stats store as it's meant to be a distinct entity.

jmarantz replied:
I think this is actually not a behavior change; we are truncating prior to this PR in thread_local_store, but not IsolatedStore.  It follows the configurable options.

jmarantz replied:
https://github.com/envoyproxy/envoy/issues/3965

mrice32 replied:
I'm not seeing where we're doing the truncation for the non-hot restart case.  Can you please point me to it?  [Here](https://github.com/envoyproxy/envoy/blob/324e628b7c85bcbeacfb49050dd7bfdcc86702b4/source/common/stats/thread_local_store.cc#L181) we seem to be passing through the full name, and I can't seem to find anywhere that we actually perform truncation (unless we're falling back in the hot restart case) in thread_local_store.

jmarantz replied:
@mrice32 : HeapStatDataAllocator::alloc() truncates in the first line of its impl:
```
HeapStatData* HeapStatDataAllocator::alloc(const std::string& name) {
  auto data = std::make_unique<HeapStatData>(truncateStatName(name));
  ...
```

mrice32 replied:
>I think this is actually not a behavior change; we are truncating prior to this PR in thread_local_store, but not IsolatedStore. It follows the configurable options.

I was referring to the code before your change.  I think we were not truncating before, and you changed the behavior such that we are.  I'm looking for where we were truncating in the non hot restart case before your change, and I can't find it.

jmarantz commented on source/common/stats/stats_impl.cc:
```diff
@@ -135,35 +135,47 @@ bool TagExtractorImpl::extractTag(const std::string& stat_name, std::vector<Tag>
+ *    int16_t flags_;
  */
-class CounterImpl : public Counter, public MetricImpl {
+template <class StatData> class CounterImpl : public Counter, public MetricImpl {
```
I think the only reasonable alternative for sharing this management code between HeapStatData and RawStatData, from a coding perspective, is making StatData a class with virtual methods. That seems suboptimal from (a) a performance perspective, adding an indirect function-call layer to a hot path and (b) size, adding the a vptr to each stat (though that's tiny compared to the string :)

The only other performant options I can think of are copying the body (not maintainable) or macros (not maintainable or debuggable).

jmarantz commented on source/common/stats/stats_impl.h:
```diff
@@ -447,17 +489,20 @@ class IsolatedStoreImpl : public Store {
                                                  std::vector<Tag>());
-        }) {}
+        }) {
+    stats_options_.max_obj_name_length_ = 1000 * 1000;
```
Actually I went down the road of having a special sentinal value or out-of-band bool to determine whether it's limitless, but it involves way more changes to the code, with associated risk.

I think 1M in terms of a stat name length is infinite enough, but I could use numeric_limits instead I suppose. I'll look at that and then write "done".

jmarantz replied:
I am always leery of going to the absolute max, and that was indeed a concern this time.  I had to divde numeric_limits by 2 to avoid an overflow in the calculation of maxNameLength().

Anyway, done.

jmarantz commented on source/common/stats/stats_impl.cc:
```diff
@@ -135,35 +135,47 @@ bool TagExtractorImpl::extractTag(const std::string& stat_name, std::vector<Tag>
- * Counter implementation that wraps a RawStatData.
+ * Counter implementation that wraps a StatData. StatData must have data members:
+ *    int64_t value_;
+ *    int64_t pending_increment_;
```
Good point -- done.

htuch commented on source/common/stats/stats_impl.cc:
```diff
@@ -135,35 +135,48 @@ bool TagExtractorImpl::extractTag(const std::string& stat_name, std::vector<Tag>
+  return existing_data;
+}
+
+void HeapStatDataAllocator::free(HeapStatData& data) {
```
Generally it makes it easier to review if we separate out code moves like this from substantive changes.

jmarantz replied:
OK...I moved HeapStatDataAllocator::free to the same relative spot where HeapRawStatDataAllocator::free used to be; though that was in a weird spot. I can move it again later.  Left a TODO.

htuch commented on source/common/stats/stats_impl.cc:
```diff
@@ -135,35 +135,48 @@ bool TagExtractorImpl::extractTag(const std::string& stat_name, std::vector<Tag>
+ *    std::atomic<int16_t> ref_count_;
  */
-class CounterImpl : public Counter, public MetricImpl {
+template <class StatData> class CounterImpl : public Counter, public MetricImpl {
```
Why do we need to use templates here vs. polymorphism? Would prefer the latter if possible for the usual template reasons (horrible compiler errors being my pet peeve).

jmarantz replied:
I'm not following your suggestion.  AFAIK polymorphism is not an alternative to templates, for code-sharing the class bodies. The most direct alternative is macros, and a functional (but less performance and memory-efficient) substitute in this case would be virtual inheritance from a base class. I think templates are the correct solution here.

htuch replied:
Polymorphism == virtual inheritance for the purpose of my comment :) OK, I'm willing to buy templates for performance reasons since stats increment etc. is actually on the hot path. I think you should add justification at the template declaration sites explicitly calling this out, since we default to "avoid templates unless absolutely necessary" in Envoy, an maybe someone comes along and undoes this if you don't ;)

jmarantz replied:
done

mrice32 replied:
Just to toss my hat in here - as I mentioned above, I agree with Harvey that the Envoy default is to avoid templates as much as possible.  I agree that this is the most performant solution for sharing code that I can think of.  However, it seems like the majority of the Envoy codebase (even portions in the hot path) use virtual function calls everywhere.  This feels a little bit like premature optimization.  We already do one virtual function call into Counter and Gauge per change since all the functions on those classes are virtual.  Do we have an idea how much an extra virtual call here will affect real-world performance?

I'm not super opinionated on this, and if no one else has issues with using templates this extensively, we can leave it as is, but it seems like this does go against general Envoy convention.

mattklein123 replied:
It's about 5 instructions for a virtual dispatch. I think it's very likely premature optimization but I didn't look into the change in very deep detail.

jmarantz replied:
Added comments in code:
```
//
// Also note that RawStatData needs to live in a shared memory block, and it's
// possible, but not obvious, that a vptr would be usable across processes. In
// any case, RawStatData is currently allocated with calloc() rather than new,
// so the usual C++ compiler assistance for setting up vptrs will not be
// available. This could be resolved with placed new.
```

mrice32 replied:
I think the way you would implement this with virtual inheritance is by wrapping a reference to the struct that is stored in shared memory in a virtual class that would be local to a single instance of Envoy.  This is similar to how CounterImpl, etc works today - just more nested. 

Anyway, I'll yield to @mattklein123 and @htuch on this since they are much more aware of what the overall Envoy conventions and perf tradeoffs are.

htuch commented on source/common/stats/stats_impl.h:
```diff
@@ -373,30 +384,60 @@ class SourceImpl : public Source {
+  // source/common/config/well_known_names.cc. This should eventually be changed to
+  // store a symbol-table of decomposed stat segments (e.g. split on "." to start) and
+  // arrays referencing those segments. The segments can be ref-counted so they
+  // no longer consume memory once deleted.
```
Either remove TODO or point to @ambuc 's latest PR here.

jmarantz replied:
Rewrote to reference the new PR.

mattklein123 commented on source/common/stats/stats_impl.h:
```diff
@@ -283,12 +273,21 @@ class MetricImpl : public virtual Metric {
- * in a shared memory block without internal pointers.
- */
-class RawStatDataAllocator : public StatDataAllocator {
+// Partially implements a StaDataAllocator, leaving alloc & free for subclasses.
```
typo StaDataAllocator

jmarantz replied:
done

mattklein123 commented on source/common/stats/stats_impl.h:
```diff
@@ -373,30 +391,56 @@ class SourceImpl : public Source {
-  RawStatData* alloc(const std::string& name) override;
-  void free(RawStatData& data) override;
+  explicit HeapStatDataAllocator(const StatsOptions& options) : StatDataAllocatorImpl(options) {}
+  // HeapStatDataAllocator() {}
```
?

jmarantz replied:
done

mrice32 commented on source/common/stats/stats_impl.cc:
```diff
@@ -135,35 +135,33 @@ bool TagExtractorImpl::extractTag(const std::string& stat_name, std::vector<Tag>
-  data->checkAndInit(name, num_bytes_to_allocate);
+HeapStatData::HeapStatData(absl::string_view key) : name_(key.data(), key.size()) {}
 
+HeapStatData* HeapStatDataAllocator::alloc(const std::string& name) {
```
I mentioned this earlier, but the comment got buried under some other changes.  I'm relatively confident that adding truncation in HeapStatDataAllocator is a user-facing change.  Previously, in non hot restart mode, we did not truncate stats.  This should be documented in the release notes.  I'm not sure it makes sense to change this behavior if we ultimately do not want non hot restart stats to be truncated as is suggested in #3965. I could be wrong about the behavior changing, but if so, please point out where they are currently being truncated on master.

jmarantz replied:
Reading again I think you are right; it's pretty subtle though. In current master, we truncate when making a RawStatData in RawStatData::truncateAndInit(), but asymmetrically, we truncate at the call-site in ThreadLocalStoreImpl::safeMakeStat() only when calling the fallback.  OK let me fix that; it's a little cleaner to just do this right.

jmarantz commented:
I restored the previous functionality, and added explicit tests for it.  I made all stat-name truncation happen in ThreadLocalStore, and added a Stat interface for determining whether an allocator requires truncation.

This changed a bunch of things -- including some changes I had previously made reverting back to the state they were in 'master'. But some interface changes, such as std::string --> absl::string_view and making the mock test allocator a separate class -- means the file-count is still 19. I think it's easiest to review this as changes from master, rather than reviewing the incremental commits.

htuch commented on source/common/stats/stats_impl.h:
```diff
@@ -283,33 +272,54 @@ class MetricImpl : public virtual Metric {
+   * @param stats_options The stat optinos, which must live longer than the allocator.
+   */
+  explicit StatDataAllocatorImpl(/*const StatsOptions& stats_options*/) {}
+  //: stats_options_(stats_options) {}
```
Nit: remove comment

jmarantz replied:
done

htuch commented on source/common/stats/stats_impl.h:
```diff
@@ -283,33 +272,54 @@ class MetricImpl : public virtual Metric {
+template <class StatData> class StatDataAllocatorImpl : public StatDataAllocator {
 public:
+  /**
+   * @param stats_options The stat optinos, which must live longer than the allocator.
```
Nit: s/optinos/options/

jmarantz replied:
done

htuch commented on source/common/stats/stats_impl.h:
```diff
@@ -283,33 +272,54 @@ class MetricImpl : public virtual Metric {
+  /**
+   * @param stats_options The stat optinos, which must live longer than the allocator.
+   */
+  explicit StatDataAllocatorImpl(/*const StatsOptions& stats_options*/) {}
```
Nit: empty arg default constructor and explicit..

jmarantz replied:
done

htuch commented on source/common/stats/thread_local_store.cc:
```diff
@@ -156,6 +156,26 @@ void ThreadLocalStoreImpl::clearScopeFromCaches(uint64_t scope_id) {
+      name = absl::string_view(name.data(), max_length);
+    }
+  }
+  return name;
```
Random thought that is unrelated to this PR (existing issue). How confident are we that two truncated stats won't conflict, e.g. if I truncated `a.b.c.d` and `a.b.c.e` to 5 characters, we get  overlap.. it seems this is double bad, since we're both truncating and conflating..

jmarantz replied:
Yeah...in the past if I wanted to truncate to 'n' I'd actually truncate to n-k, and use a k-character base64 hash of the untruncated string to disambiguate.  That would still be pretty awful for stats as you'd have no idea which one you were looking at. Basically, truncation is bad and should be avoided.

htuch commented on source/common/stats/thread_local_store.cc:
```diff
@@ -178,18 +198,13 @@ StatType& ThreadLocalStoreImpl::ScopeImpl::safeMakeStat(
   if (!central_ref) {
     std::vector<Tag> tags;
     std::string tag_extracted_name = parent_.getTagsForName(name, tags);
```
Can you add a comment explaining why it's not necessary to use the truncated name for tag extraction?

htuch commented on test/test_common/utility.h:
```diff
@@ -369,6 +371,27 @@ class TestAllocator : public RawStatDataAllocator {
+
+class MockedTestAllocator : public RawStatDataAllocator {
+public:
+  MockedTestAllocator(const StatsOptions& stats_options) : alloc_(stats_options) {
```
Nit: constructor/destructor for mocks should live in `.cc`. See https://github.com/google/googlemock/blob/master/googlemock/docs/v1_6/CookBook.md#making-the-compilation-faster

jmarantz replied:
done

mattklein123 commented on source/common/stats/stats_impl.h:
```diff
@@ -283,33 +272,53 @@ class MetricImpl : public virtual Metric {
+  /**
+   * @param stats_options The stat options, which must live longer than the allocator.
+   */
+  StatDataAllocatorImpl() {}
```
nit: Is this default constructor needed? Doc comment seems out of date also.

jmarantz replied:
oops; not needed; in an earlier rev I needed a 1-arg constructor and removing the arg, I forgot to remove the whole thing.

mattklein123 commented on source/common/stats/thread_local_store.cc:
```diff
@@ -156,6 +156,26 @@ void ThreadLocalStoreImpl::clearScopeFromCaches(uint64_t scope_id) {
+
+    // Note that the heap-allocator does not truncate itself; we have to
+    // truncate here if we are using heap-allocation as a fallback due to an
+    // exahusted shared-memroy block
```
typo memroy

jmarantz replied:
done

mattklein123 commented on source/server/hot_restart_impl.cc:
```diff
@@ -141,14 +141,16 @@ HotRestartImpl::HotRestartImpl(Options& options)
-    key.remove_suffix(key.size() - options_.statsOptions().maxNameLength());
+  // The name is truncated in ThreadLocalStore before this is called.
+  size_t max_length = options_.statsOptions().maxNameLength();
+  if (name.length() > max_length) {
```
Can this happen then? Should this just be an assert? The comment is a bit confusing.

jmarantz replied:
It should not happen in production. It can happen in a unit test.  It could be an ASSERT, but then it can't be tested.  It could be a RELEASE_ASSERT, and be tested with a DEATH_TEST.

I'm happy to go with whatever is the best practice of those for the project.

I'll try to clarify the comment.

mattklein123 replied:
I would probably just do a normal ASSERT as this seems like a programming error. If you feel strongly that it should be a RELEASE_ASSERT w/ a test that's fine also.

mrice32 commented on source/server/hot_restart_impl.h:
```diff
@@ -139,8 +139,9 @@ class HotRestartImpl : public HotRestart,
-  Stats::RawStatData* alloc(const std::string& name) override;
+  Stats::RawStatData* alloc(absl::string_view name) override;
   void free(Stats::RawStatData& data) override;
+  bool requiresBoundedStatNameSize() const override { return true; }
```
Why is this overridden here and in `RawStatDataAllocator`?  Also why even have the `RawStatDataAllocator` intermediate class?

jmarantz replied:
Good catch - I didn't need this extra impl of requiresBoundedStatNameSize here, but I do need RawStatDataAllocator as it is used as a base class for test allocators as well as this.

mrice32 commented on source/common/stats/stats_impl.cc:
```diff
@@ -319,47 +320,28 @@ TagProducerImpl::addDefaultExtractors(const envoy::config::metrics::v2::StatsCon
 }
 
-void HeapRawStatDataAllocator::free(RawStatData& data) {
+// TODO(jmarantz): move this below HeapStatDataAllocator::alloc.
```
?  Seems like a strange TODO?

jmarantz replied:
Normally it makes sense to group alloc/free together of a single class, but these got separated in a prior PR; leaving this here now for easier diffs.  Will fix later in a follow-up.

mrice32 replied:
I see. It seemed so trivial, I was a little confused as to why you didn't just do it now. Easier review is a good reason. That makes sense.

mrice32 commented on source/common/stats/stats_impl.cc:
```diff
@@ -135,35 +135,36 @@ bool TagExtractorImpl::extractTag(const std::string& stat_name, std::vector<Tag>
+HeapStatData::HeapStatData(absl::string_view key) : name_(key.data(), key.size()) {}
 
+HeapStatData* HeapStatDataAllocator::alloc(absl::string_view name) {
+  // Do not truncate here for non-hot-restart case; truncate at the call-site in
```
This is probably a weird nit, but I think this comment is adding confusion.  If we want to delegate truncation to the callsite, I think we can just make the comment (or something similar) since the truncation logic isn't really relevant here:
```
// Any expected truncation of name is done at the callsite. No truncation is
// required to use this allocator.
```

jmarantz replied:
done

mrice32 commented on source/common/stats/stats_impl.h:
```diff
@@ -373,30 +377,58 @@ class SourceImpl : public Source {
-  typedef std::unordered_set<RawStatData*, RawStatDataHash_, RawStatDataCompare_> StringRawDataSet;
 
-  // An unordered set of RawStatData pointers which keys off the key()
+  // TODO(jmarantz): See https://github.com/envoyproxy/envoy/pull/3927 which can
```
Sounds like more of a TODO for @ambuc :)

jmarantz replied:
In my experience, TODO's are generally signed by the author of the TODO, not necessarily by the person who is expected to do it.

ambuc replied:
Let's leave a comment linking to issue https://github.com/envoyproxy/envoy/issues/3585 which will address this in #3927, hopefully.

mrice32 commented on source/common/stats/thread_local_store.cc:
```diff
@@ -156,6 +156,26 @@ void ThreadLocalStoreImpl::clearScopeFromCaches(uint64_t scope_id) {
+absl::string_view ThreadLocalStoreImpl::truncateStatNameIfNeeded(absl::string_view name) {
+  // If the main allocator requires stat name truncation, warn and truncate, before
+  // attempting to allocate.
+  if (alloc_.requiresBoundedStatNameSize()) {
```
nit: as I mentioned in person, I think the encapsulation would be cleaner if the allocators were constructed with their truncation settings and then they could handle any necessary truncation themselves (the fallback heap allocator would be set up to truncate, while the non-hot restart heap allocator would be set up to not truncate).  Here, it seems like we're half delegating that responsibility by giving them a method to communicate whether they require truncation, but not giving them the responsibility to read the settings and do that truncation within `alloc()`.

However, I think we should just try to get this PR in as is rather than making more interface changes, so this seems fine for now.

jmarantz replied:
Yeah I think that's a follow-up change.

mrice32 commented on source/common/stats/stats_impl.h:
```diff
@@ -373,30 +377,58 @@ class SourceImpl : public Source {
+  HeapStatData* alloc(absl::string_view name) override;
+  void free(HeapStatData& data) override;
+
+  // StatsDataAllocator
```
nit typo: `StatDataAllocator`

jmarantz replied:
done

mrice32 replied:
This typo seems to still be here :)

mrice32 commented on source/common/stats/stats_impl.h:
```diff
@@ -283,33 +272,48 @@ class MetricImpl : public virtual Metric {
+
+class RawStatDataAllocator : public StatDataAllocatorImpl<RawStatData> {
+public:
+  // StatsDataAllocator
```
nit: typo here, too: `StatDataAllocator`
